qutip/core/data/matmul.pyx

```diff
@@ -192,6 +192,58 @@ cpdef CSR matmul_csr(CSR left, CSR right, double complex scale=1, CSR out=None):
     mem.PyMem_Free(nxt)
     return out
 
+@cython.boundscheck(False)
+@cython.wraparound(False)
+@cython.profile(False)
+@cython.linetrace(False)
+cdef void _imatmul_coo_dense_dense(COO left, Dense right,
+                                   double complex scale, Dense out) nogil:
+    cdef idxint ptr_l, ptr_r, ptr_out, col
+    cdef idxint right_row_stride = 1 if right.fortran else right.shape[1]
+    cdef idxint right_col_stride = right.shape[0] if right.fortran else 1
+
+    cdef idxint out_row_stride = 1 if out.fortran else out.shape[1]
+    cdef idxint out_col_stride = out.shape[0] if out.fortran else 1
+
+    for ptr_l in range(coo.nnz(left)):
+        for col in range(right.shape[1]):
+            ptr_out = left.row_index[ptr_l] * out_row_stride + col * out_col_stride
+            ptr_r = left.col_index[ptr_l] * right_row_stride + col * right_col_stride
+            out.data[ptr_out] += scale *  left.data[ptr_l] * right.data[ptr_r]
+
+cpdef Dense matmul_coo_dense_dense(COO left, Dense right,
+                                   double complex scale=1, Dense out=None):
+    """
+    Perform the operation
+        ``out := scale * (left @ right) + out``
+    where `left`, `right` and `out` are matrices.  `scale` is a complex scalar,
+    defaulting to 1.
+
+    If `out` is not given, it will be allocated as if it were a zero matrix.
+    """
+    _check_shape(left, right, out)
+    if out is None:
+        out = dense.zeros(left.shape[0], right.shape[1], right.fortran)
+        
+    _imatmul_coo_dense_dense(left, right, scale, out)
+    return out
+
+
+cpdef COO matmul_coo(COO left, COO right,
+                                   double complex scale=1):
+    _check_shape(left, right)
+    cdef COO out = coo.empty(left.shape[0], right.shape[1], 1)
+    cdef idxint ptr_l, ptr_r
+    for ptr_l in range(coo.nnz(left)):
+        for ptr_r in range(coo.nnz(right)):
+            if left.col_index[ptr_l] == right.row_index[ptr_r]:
+                if out._nnz >= out.size:
+                    out = coo.expand(out, 2 * out.size)
+                out.data[out._nnz] = scale * left.data[ptr_l] * right.data[ptr_r]
+                out.row_index[out._nnz] = left.row_index[ptr_l]
+                out.col_index[out._nnz] = right.col_index[ptr_r]
+                out._nnz += 1
```

Comment:
While I don't dislike the use of `expand` here, you can easily compute the output's nnz since you allow duplicate entries.